### PR TITLE
Use qemu over vagrant for testing

### PR DESCRIPTION
**Pull request by alexdavid**
This replaces the current vagrant/virtualbox setup with a call to qemu to make the test dependencies of this repo be completely deterministic within nix.

### CODE REVIEW

**Comment by alexdavid** [comment]
There still seems to be a size problem where the downloaded ubuntu image is only 512MB, but this isn't enough to run the hello package in the test :cry: 

We can of-course re-size the image (but that alone isn't enough since we also need to boot the VM and resize the filesystem partition within since `qemu-img resize` only resizes the raw block device). I wonder if maybe there's a better way to resolve this.

**Comment by alexdavid** [comment]
Turns out I was mistaken, the qcow2 image is only 512MB, but the actual size is 2G. Also, it turns out the cloud image automatically resizes the filesystem when resizing the backing image, so that's all we needed to do!

File: flake.nix [hunk 1]
```diff
@@ -88,6 +88,72 @@
...
+                  ];
+                };
+                cloud-init-img = pkgs.runCommand "user-data-img" {}
+                  "${pkgs.cloud-utils}/bin/cloud-localds $out ${pkgs.writeTextFile {
+                    name = "userdata";
+                    text = "#cloud-config\n" + builtins.toJSON cloudcfg;
+                  }}";
+              in
+                builtins.toString (pkgs.writeScript "boot-vm" ''
+                  ROOT_IMG=$(mktemp)
```
Comment (soenkehahn) [inline flake.nix]:
```suggestion
                builtins.toString (pkgs.writeScript "boot-vm" ''
                  set -eux
                  
                  ROOT_IMG=$(mktemp)
```

File: justfile [hunk 2]
```diff
@@ -7,10 +7,25 @@ test-run-in-vagrant: test-setup
   nix run -L .#testFileServer &
   SERVER_PID=$!
   trap 'kill $SERVER_PID' SIGINT SIGTERM EXIT
-  vagrant ssh -c "echo ; echo ; echo ; sh <(curl --tlsv1.2 -sSf http://192.168.56.1:8000/install.sh)"
+  while ! nc -z localhost 8000; do
+    sleep 0.1
+  done
+  nix run .#sshVm -- "echo ; echo ; echo ; sh <(curl --tlsv1.2 -sSf http://10.0.2.2:8000/install.sh)"
```
Comment (soenkehahn) [inline justfile]:
```suggestion
  nix run -L .#sshVm -- "echo ; echo ; echo ; sh <(curl --tlsv1.2 -sSf http://10.0.2.2:8000/install.sh)"
```

File: justfile [hunk 3]
```diff
@@ -7,10 +7,25 @@ test-run-in-vagrant: test-setup
...
+
+test-ssh: test-setup
+  nix run .#sshVm
```
Comment (soenkehahn) [inline justfile]:
```suggestion
  nix run -L .#sshVm
```

File: justfile [hunk 4]
```diff
@@ -7,10 +7,25 @@ test-run-in-vagrant: test-setup
...
+  #!/usr/bin/env bash
+
+  set -eu
+
+  if lsof -i tcp:2222; then
+    echo Port 2222 is already in use...
+    echo Assuming server is already running
+    exit 0
+  fi
+  nix run .#bootVm
```
Comment (soenkehahn) [inline justfile]:
```suggestion
  nix run -L .#bootVm
```

File: flake.nix [hunk 5]
```diff
@@ -88,6 +88,72 @@
...
+                  users = [{
+                    name = "test";
+                    plain_text_passwd = "test";
+                    lock_passwd = false;
+                    sudo = "ALL=(ALL) ALL";
+                    shell = "/bin/bash";
+                  }];
+                  runcmd = [
+                    # Attempt to briefly connect to port 2223 on the host to
+                    # notify the boot-vm script know that we have fully booted
```
Comment (soenkehahn) [inline flake.nix]:
```suggestion
                    # notify the boot-vm script that we have fully booted
```

File: flake.nix [hunk 6]
```diff
@@ -88,6 +88,72 @@
...
+                    -device intel-iommu \
+                    -drive file=$ROOT_IMG,format=qcow2 \
+                    -drive file=$CLOUD_INIT_IMG,format=raw \
+                    -nographic \
+                    -device e1000,netdev=net0 \
+                    -netdev user,id=net0,hostfwd=tcp::2222-:22 \
+                    &
+                  # Wait for the runcmd to call us back to let us know that we are fully booted.
+                  # (See runcmd in the flake file under cloudcfg)
+                  nc -l 2223
```
Comment (soenkehahn) [inline flake.nix]: Any reason to use `-v` in one of the `nc` commands but not the other?

**Review by soenkehahn — APPROVED** [review]
Left a bunch of comments, apart from them, LGTM!

**Comment by soenkehahn** [comment]
@alexdavid: One more thing: We currently don't have any infrastructure to help with formatting nix code in this repo. But maybe it makes sense to still run `nixpkgs-fmt` on the flake file? Up to you!

File: flake.nix (hunk 6 above)
Comment (alexdavid) [reply flake.nix]: I'm just used to appending a `-v` for connecting `nc` calls and not on listening ones, but you bring up a good point that having it on both is more helpful for debugging.

**Pull request merged**